packages/concerto-tools/lib/codegen/fromcto/openapi/openapivisitor.js

```diff
@@ -0,0 +1,287 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const pluralize = require('pluralize');
+
+const JSONSchemaVisitor = require('../jsonschema/jsonschemavisitor');
+
+/**
+ * Convert the contents of a ModelManager
+ * to an Open API 3.0.2 specification document, where
+ * each concept declaration with a @resource decorator
+ * becomes a RESTful resource addressable via a path.
+ *
+ * The JSON schema for the types is included in the
+ * Open API document and is used to validate request and
+ * response bodies.
+ *
+ * @private
+ * @class
+ */
+class OpenApiVisitor {
+    /**
+     * Visitor design pattern
+     * @param {Object} thing - the object being visited
+     * @param {Object} parameters  - the parameters
+     * @param {string} [parameters.openApiTitle] - the title property
+     * of the Open API specification
+     * @param {string} [parameters.openApiVersion] - the version property
+     * of the Open API specification
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visit(thing, parameters) {
+        if (thing.isModelManager?.()) {
+            return this.visitModelManager(thing, parameters);
+        } else if (thing.isModelFile?.()) {
+            return this.visitModelFile(thing, parameters);
+        } else if (thing.isClassDeclaration?.()) {
+            return this.visitClassDeclaration(thing, parameters);
+        } else {
+            throw new Error('Unrecognised ' + JSON.stringify(thing));
+        }
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ModelManager} modelManager - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitModelManager(modelManager, parameters) {
+        // get all the types
+        const visitor = new JSONSchemaVisitor();
+        const childParameters = {
+            refRoot: '#/components/schemas',
+        };
+        const jsonSchema = modelManager.accept(visitor, childParameters);
+
+        // Visit all of the files in the model manager.
+        const title = parameters?.openApiTitle
+            ? parameters.openApiTitle
+            : 'Generated Open API from Concerto Models';
+        const version = parameters?.openApiVersion
+            ? parameters.openApiVersion
+            : '1.0.0';
+
+        let result = {
+            openapi: '3.0.2',
+            servers: parameters?.openApiServers ? parameters.openApiServers : [],
```

Comment:
Could be reduced to `parameters?.openApiServers ?? []` but perhaps our level of JS doesn't allow for it.

Reply:
Thanks. I've created a follow up PR for that.